### ISSUE TITLE
Error in Search, when use Forbid within tear in Knowledge Box

### ISSUE THREAD

**Issue by barrynuqoba** [cmu-phil/tetrad#1541]
Hi,

I use TETRAD 7.2.2 to discover a causal graph on 16 variables. I use a Knowledge box consisting of 3 tiers: 9 variables (first tier - Forbid Within Tier), 6 variables (second tier), 1 variable (third tier). When I ran GRaSP algorithm using a Search box, it raised an error mentioning, "Stopped with error: The initial sorting procedure could not find a permutation consistent with that knowledge". When I remove the Forbid Within Tier in the first tier, the Search box can yield a causal graph result without error.

Moreover, previously I used TETRAD 7.1.0 for the same cases as above and found no error.

Should I use TETRAD 7.1.0 instead of the latest version (7.2.2)?

Thank you very much for your incredible work on TETRAD. It is beneficial for my study.

Best regards,
Barry

**Comment by cg09** [comment]
Try 7.1.3

**Comment by barrynuqoba** [comment]
Thank you, Clark. I will try it.

**Comment by cg09** [comment]
Let me know what happens, please. We have found (and Joe has corrected)
other problems with 7.2. I found 7.1.3 pretty smooth sailing.

**Comment by barrynuqoba** [comment]
Hi Clark. I just tried on 7.1.3 and 7.2.0, both can run smoothly and no error.
Thank you for your clarifications, I will use 7.1.3 or 7.2.0 then.

**Comment by barrynuqoba** [comment]
Hi, Clark. I want to inform you that after I double checked on the causal graph result, I found that TETRAD 7.2.0 also has issues on the Knowledge Box. The results shows variable A (in tier 3) can cause variable B (in tier 2).

**Comment by cg09** [comment]
Thanks.

**Comment by jdramsey** [comment]
@barrynuqoba No, I just added that exception, because someone had reported another problem to me (that I fixed). So going to an earlier version isn't going to help.

The way I'd done knowledge for GRaSP was, I tried to sort the varaibles into an order acceptable to knowledge, and then after each transformation I'd see if the variables were still consistent with knowledge. If the initial sort is incompatible with knowledge, that procedure won't work.

Let me think a second.

**Comment by jdramsey** [comment]
Did you place any constraints on the knowledge, such forbidding within tier?

**Comment by jdramsey** [comment]
Oh wait, you did. It's possible there just isn't a DAG consistent with the "forbid within tier" knowledge. The problem is GRaSP, unlike PC, say, is actually building a DAG each time, so if it can't find a specific DAG, it may not succeed.

**Comment by jdramsey** [comment]
There's another way to do knowledge that @bja43 had proposed that may be more successful in these cases. We may try that for the next version. That involves forbidding and requiring edges _at the step where the DAGs are built_, instead of looking to see after the DAG is built whether it satisfies knowledge. We may try that for the next version.

In any case, I just made a model with 20 nodes and put it into roughly the same counts of tiers that you did and the sort succeeded. And when I did the forbid within tier thing, it failed, as yours. My knowledge of course was all wrong on purpose, so it's possible it didn't succeed for that reason.

Let me try correct tiers. Nope, even with correct tiers ther are edges in the DAG in Tier 1. So it may just not be able to find a DAG where there are not edges in Tier 1.

That is a problem for some procedures.

**Comment by jdramsey** [comment]
Huh, even in case where the actually _were_ no edge between variables in Tier 1, it failed to find an initial sort consistent with knowledge. So the sorting procedure isn't working for these 3-tiered case. Let me think.

**Comment by jdramsey** [comment]
Oh I see the problem. Hold on. Let me see if I can fix this.

**Comment by barrynuqoba** [comment]
Hi Joe.

Thank you very much for your kind help. I tried to use TETRAD 7.1.3 as suggested by Clark, and GRaSP can work well in 3 and 4 tiers (for cross-sectional and panel data), also when I included "Forbid Within Tier" and Forbidden Edges.

Best regards,
Barry

**Comment by jdramsey** [comment]
I think that may solve that particular problem, but it doesn't solve the problem in general, because there may still be edges in Tier 1 when the DAGs are built, in which case the error message is correct. So it may not be successfully forbidding edges in Tier 1. You can check. Also, double check to make sure it's getting the tiers right in the output even if you don't select "forbid within tier"; someone else reported a bug for that.

Anyway, I'll consider putting my fix in the next version.

**Comment by jdramsey** [comment]
Actually could i get you to try my version? I'm very curious. In this version I actually build the DAG, and if the Tier 1 contains X1 and X2, say, and you select Forbid Within Tier for Tier 1, instead of looking to see if forbid(X1, X2) (which is true), it looks to see if forbid(X1, X2) and there's an edge from X1 to X2 (fixing that bug).

Are you using the Tetrad interface? I could send you a revised jar to try.

**Comment by barrynuqoba** [comment]
Sure, Joe. I will try your version. Yes, I use the TETRAD GUI. Please send me the revised jar. Thanks.

**Comment by jdramsey** [comment]
Give me a few minutes...I just added some perhaps helpful output.

**Comment by barrynuqoba** [comment]
Hi Joe,

I have double-checked my result for panel data using TETRAD 7.1.3. I used four tiers. The tiers are as follows: tier 1 (3 variables, Forbid Within Tier), tier 2 (13 variables), tier 3 (13 variables), and tier 4 (13 variables). The causal graph result shows no edges between variables in tier 1. I think 7.1.3 already works well.

However, the tier issues appear in 7.2.0, as variables in a tier can cause variables in the previous tier.

I hope it clarifies.

**Comment by jdramsey** [comment]
Well if you're having success, I'll work on the issue separately.

If you are willing to try the version I just made here's the link.

https://s01.oss.sonatype.org/content/repositories/snapshots/io/github/cmu-phil/tetrad-gui/7.3.3_20230302.065037-1/tetrad-gui-7.3.3_20230302.065037-1-launch.jar

**Comment by barrynuqoba** [comment]
Thank you very much, Joe. I will try to use the version you just made.

**Comment by jdramsey** [comment]
Thanks; I'll keep working on the issue as well. Hopefully by the next released version I'll have it solved. Knowledge is a bit tricky, always, and the PC way of doing it doesn't work for GRaSP.

**Comment by jdramsey** [comment]
@barrynuqoba Turns out @bja43 will include some fixes to knowledge for the permutation algorithms in the next version, which should be out in another week or so...

**Comment by barrynuqoba** [comment]
Sounds great, @jdramsey! Thank you very much for your kind updates. Can't wait to try the new version for my experiments.

**Comment by jdramsey** [comment]
Hi @barrynuqoba, We released version 7.3.0, which solves your problem. For BOSS, we still find if you use knowledge, there is a problem in the BES step, which we've reported here:

https://github.com/cmu-phil/tetrad/issues/1565

Hopefully, this will get fixed soon. But knowledge for GRaSP and SP are both flawless and knowledge for BOSS is working, except for cases where it falsely judges a graph as "circular" (so you can tell when the problem happens). This is rare; we can only get it to happen in BFCI.

The use of knowledge is interesting. If you provide knowledge "tiers," BOSS/GRaSP/SP will analyze each tier separately, so you can analyze larger problems this way if you have temporal tiers (for instance, when you're analyzing time series. So for SP, which is usually limited to about ten variables (here, 12 or so), if you provide knowledge tiers, you can extend that to arbitrary numbers of tiers of 10 variables. The problem will finish in a reasonable time. This is thanks to @bja43.

**Comment by jdramsey** [comment]
Let us know if this solves your problem; whenever you think it's solves you can close the issue. :)

**Comment by barrynuqoba** [comment]
Hi, @jdramsey.  Thank you very much for your fantastic news. Also, thank you to @bja43 and all TETRAD teams for their excellent work. I will use the newly released TETRAD (7.3.0) for my experiments over the weekend. I will let you know if it can work well, and I will close the issue afterward. :)

**Comment by barrynuqoba** [comment]
Hi @jdramsey , I have tried the new TETRAD (7.3.0) version, and it works great! Thank you very much to you, @bja43 , and all TETRAD teams. :)

**Comment by jdramsey** [comment]
Great, thanks!
